vnext/Microsoft.ReactNative.SharedManaged/ReactMethodInfo.cs

```diff
@@ -2,304 +2,204 @@
-      ParameterExpression resolveParameter = Expression.Parameter(typeof(MethodResultCallback), "resolve");
-      ParameterExpression rejectParameter = Expression.Parameter(typeof(MethodResultCallback), "reject");
-
-      // Input variables to read from inputReader
-      ParameterExpression[] inputVariables = new ParameterExpression[parameters.Length];
-      for (int i = 0; i < inputVariables.Length; ++i)
-      {
-        inputVariables[i] = Expression.Variable(parameters[i].ParameterType, parameters[i].Name);
+        return false;
       }
 
-      // Statements of the generated lambda
-      List<Expression> statements = new List<Expression>();
+      bool isCallback(Type type) => type != null && typeof(Delegate).IsAssignableFrom(type);
 
-      // Generate code to read input variables from the inputReader
-      var callReadNext = Expression.Call(inputReaderParameter, typeof(IJSValueReader).GetMethod("ReadNext"));
-      statements.Add(callReadNext);
-      foreach (ParameterExpression variable in inputVariables)
+      if (returnType != typeof(void))
       {
-        statements.Add(callReadNext);
-        statements.Add(Expression.Call(null, JSValueReader.GetReadValueMethod(variable.Type), inputReaderParameter, variable));
+        MethodReturnType = MethodReturnType.Callback;
+        createMethod = () => MakeReturningMethod(methodInfo, returnType, parameters);
       }
-
-      // Generate code to call the method
-      statements.Add(Expression.Call(
-          Expression.Convert(moduleParameter, methodInfo.DeclaringType),
-          methodInfo,
-          inputVariables));
-
-      // Generate the lambda to return
-      var lambda = Expression.Lambda<ReactMethodImpl>(
-          Expression.Block(inputVariables, statements),
-          moduleParameter, inputReaderParameter, outputWriterParameter, resolveParameter, rejectParameter);
-
-      // Compile and return the lambda
-      return lambda.Compile();
-    }
-
-    private ReactMethodImpl MakeCallbackMethod(ReflectionMethodInfo methodInfo, ParameterInfo[] parameters)
-    {
-      // The last variable is a delegate
-
-      // Input parameters for generated lambda
-      ParameterExpression moduleParameter = Expression.Parameter(typeof(object), "module");
-      ParameterExpression inputReaderParameter = Expression.Parameter(typeof(IJSValueReader), "inputReader");
-      ParameterExpression outputWriterParameter = Expression.Parameter(typeof(IJSValueWriter), "outputWriter");
-      ParameterExpression resolveParameter = Expression.Parameter(typeof(MethodResultCallback), "resolve");
-      ParameterExpression rejectParameter = Expression.Parameter(typeof(MethodResultCallback), "reject");
-
-      // Input variables to read from inputReader
-      ParameterExpression[] inputVariables = new ParameterExpression[parameters.Length - 1];
-      for (int i = 0; i < inputVariables.Length; ++i)
+      else if (isPromise(lastParameterType))
```

Comment:
>lastParameterType [](start = 25, length = 17)

Having a promise as an explicit parameter feels a bit weird to me. How difficult would it be to be able to use stock c# functions returning something like `Task<ReactResult<T>>`? That would have the benefit of forcing methods to return something.

Reply:
Good idea! We can consider adding it in future. We use similar approach in our Reka implementation by returning Mso::Future. Though, I find the creation and managing the return result a little bit more complex in that case. While Task<T> enforces developers to return something, please note that developers write this method themselves and returning Task<T> instead of receiving IReactPromise<T> would also require them to create it and set instead of just setting values. Plus the whole business of scheduling task on another thread by default can also end up to be a non-trivial task.